subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/files/SourcepathIgnoringJavaFileManager.java

```diff
@@ -41,7 +44,22 @@ public SourcepathIgnoringJavaFileManager(JavaFileManager fileManager) {
      * This implementation of hasLocation() pretends that the JavaFileManager
      * has no concept of a source path.
      */
+    @Override
     public boolean hasLocation(Location location) {
         return !location.equals(StandardLocation.SOURCE_PATH) && fileManager.hasLocation(location);
     }
+
+    /**
+     * If we are pretending that we don't have a sourcepath, the compiler will
+     * look on the classpath for sources. Since we don't want to bring in any
+     * sources implicitly from the classpath, we have to ignore source files
+     * found on the classpath.
+     */
+    @Override
+    public Iterable<JavaFileObject> list(Location location, String packageName, Set<JavaFileObject.Kind> kinds, boolean recurse) throws IOException {
```

Comment:
:x: Probably needs a test case/integration test: I assume that if you added this, that's because you discovered a bug, that is not covered by a test case.